core/services/ocr2/plugins/ccip/ccipexec/batching.go

```
@@ -3,19 +3,335 @@ package ccipexec
+	sourceTokenPricesUSD       map[cciptypes.Address]*big.Int
+	destTokenPricesUSD         map[cciptypes.Address]*big.Int
+	gasPrice                   *big.Int
+	sourceToDestToken          map[cciptypes.Address]cciptypes.Address
+	aggregateTokenLimit        *big.Int
+	tokenDataRemainingDuration time.Duration
+	tokenDataWorker            tokendata.Worker
+	gasPriceEstimator          prices.GasPriceEstimatorExec
+	destWrappedNative          cciptypes.Address
+	offchainConfig             cciptypes.ExecOffchainConfig
+}
+
+type BaseBatchingStrategy struct{}
+
+type BatchingStrategy interface {
+	BuildBatch(ctx context.Context, batchCtx *BatchContext) ([]ccip.ObservedMessage, []messageExecStatus)
+}
+
+type BestEffortBatchingStrategy struct {
+	BaseBatchingStrategy
+}
+
+type ZKOverflowBatchingStrategy struct {
+	BaseBatchingStrategy
+	statuschecker statuschecker.CCIPTransactionStatusChecker
+}
+
+// BestEffortBatchingStrategy is a batching strategy that tries to batch as many messages as possible (up to certain limits).
+func (s *BestEffortBatchingStrategy) BuildBatch(
+	ctx context.Context,
+	batchCtx *BatchContext,
+) ([]ccip.ObservedMessage, []messageExecStatus) {
+	inflightAggregateValue, err := getInflightAggregateRateLimit(batchCtx.lggr, batchCtx.inflight, batchCtx.destTokenPricesUSD, batchCtx.sourceToDestToken)
+	if err != nil {
+		batchCtx.lggr.Errorw("Unexpected error computing inflight values", "err", err)
+		return []ccip.ObservedMessage{}, nil
+	}
+
+	batchBuilder := newBatchBuildContainer(len(batchCtx.report.sendRequestsWithMeta))
+	for _, msg := range batchCtx.report.sendRequestsWithMeta {
+		msgLggr := batchCtx.lggr.With("messageID", hexutil.Encode(msg.MessageID[:]), "seqNr", msg.SequenceNumber)
+		shouldAdd, status, messageMaxGas, tokenData, msgValue, err := s.performCommonChecks(ctx, batchCtx, inflightAggregateValue, msg, msgLggr)
+
+		if err != nil {
+			return []ccip.ObservedMessage{}, []messageExecStatus{}
+		}
```

This error seems to be swallowed, why not bubble it up?

---

In the current logic, we have one occurrence where we are not skipping the item but directly stopping the batching for the report. When getting an error from `gasPriceEstimator`, we are simply returning an empty result (see [code](https://github.com/smartcontractkit/ccip/blob/ccip-develop/core/services/ocr2/plugins/ccip/ccipexec/ocr2.go#L387)).

I decided to be as close as possible to the current logic as I wasn't sure why we decided that. Most probably we are considering this error as fatal and we want to stop the batching logic here for this report?

---

But it's a surprising behavior. I really don't see why we are returning empty instead of just skipping the item

---

Yeah, I do agree, current behavior is brittle, we shouldn't swallow an error and instead skip a message and log its state as "can't estimate USD cost" or sth like this

@0xnogo can you please create a ticket for that? It could be nice onboarding task